lib/src/bitarray.dart

```diff
@@ -0,0 +1,317 @@
+/// * instance methods (mutate [this]): [BitArray.not], [BitArray.or], [BitArray.and], [BitArray.xor];
+/// * producing new instance of [BitArray]:
+///     * logical operators: [BitArray.~], [BitArray.|], [BitArray.&], [BitArray.^].
+///     * logical gates (as static methods): [BitArray.notGate], [BitArray.orGate], [BitArray.andGate], [BitArray.xorGate].
+/// All arguments for not-unary logical operations must have the same [length] or [ArgumentError] will be thrown.
+class BitArray with ListMixin<bool>, NonGrowableListMixin<bool> {
+  int _length;
+
+  Uint32List _data;
+
+  /// Mask to set tail bits to zero.
+  int _tailMask;
+
+  static const List<int> _numOfBitsLookup = [
+    0, 1, 1, 2, 1, 2, 2, 3, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    1, 2, 2, 3, 2, 3, 3, 4, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    2, 3, 3, 4, 3, 4, 4, 5, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    3, 4, 4, 5, 4, 5, 5, 6, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    4, 5, 5, 6, 5, 6, 6, 7, //
+    5, 6, 6, 7, 6, 7, 7, 8
+  ];
+
+  /// Creates a [BitArray] with [length].
+  ///
+  /// All initial values are [false], set [filled] to [true] to make them [true].
+  BitArray(int length, [bool filled = false]) {
+    ArgumentError.checkNotNull(length);
+    RangeError.checkNotNegative(length, 'length');
+
+    _length = length;
+    _data = Uint32List((length >> 5) + 1);
+
+    _tailMask = 0xFFFFFFFF >> (_data.length * 32 - _length);
+
+    if (filled) {
+      _setDataWords((_) => -1);
+    }
+  }
+
+  /// Generates a [BitArray] filled with values by [generator].
+  factory BitArray.generate(int length, bool Function(int index) generator) {
+    ArgumentError.checkNotNull(length);
+    RangeError.checkNotNegative(length, 'length');
+
+    ArgumentError.checkNotNull(generator);
+
+    var result = BitArray(length);
+    for (var pos = 0; pos < length; ++pos) {
+      result[pos] = generator(pos);
+    }
+    return result;
+  }
+
+  /// Replaces integer values in [_data] by applying [func] to each word.
+  void _transformDataWords(int Function(int word, int wordIndex) func) {
+    for (var wordIndex = 0; wordIndex < _data.length; ++wordIndex) {
+      _data[wordIndex] = func(_data[wordIndex], wordIndex);
+    }
+    _clearTailBits();
```

Comment:
This function updates by doing a function call per word. Most operations are so simple that inlining the loop and operation will still be almost a one-liner, and it will probably be quite a lot more efficient (unless the compiler manages to inline this function everywhere).

This is general reusable library code, it should be very efficient as a base-line because users who depend on it cannot change its performance.

Reply:
dart2js and DDC will not inline this code.

Reply:
Will be removed. This function calls will be replaced with loops.